#161342860 Implement role-based access control functionality

victor-nweze-deel opened the pull request:
#### What does this PR do?
- *Implement role-based access control functionality*

#### Description of Task to be completed?
- Create a permissions config file
- Create a middleware to verify user rights
- Include middleware to a protected route
- Write integration test for the new middleware

#### How should this be manually tested?
- Git clone this repo
- Cd into the `haven-ah-backend` folder
- Start postman and fire a POST request to `localhost:500/api/v1/articles` , say, with both valid and invalid credentials

#### Screenshots (if appropriate)
<img width="1209" alt="screen shot 2018-11-12 at 8 12 42 am" src="https://user-images.githubusercontent.com/41582565/48332092-dd576880-e652-11e8-89e3-20a61b76be71.png">
<img width="1207" alt="screen shot 2018-11-12 at 8 13 20 am" src="https://user-images.githubusercontent.com/41582565/48332096-e0525900-e652-11e8-9ae2-9c87ca4c8a31.png">


#### What are the relevant pivotal tracker stories?
[#161342860](https://www.pivotaltracker.com/story/show/161342860)

TheoOkafor commented on server/middlewares/checkPermissions.js:
```diff
@@ -0,0 +1,39 @@
+    return next();
+  }
+
+  if (!(
```
Nice one. I like this part. 👍

TheoOkafor reviewed (commented):
Nice work @ProfJigsaw

seyi-adeleke commented on server/middlewares/checkPermissions.js:
```diff
@@ -0,0 +1,39 @@
+
+  const { email, role } = user.dataValues;
+
+  if (role === 'admin' && email === process.env.SUPER_ADMIN_EMAIL) {
```
why isnt this just `role === 'admin'`

victor-nweze-deel replied:
We decided that a user could be upgraded to an admin but this doesnt give him superadmin privileges like deleting other admins or modifying the role of a superadmin. I am making a check for the superadmin (using the email is as good as using any other primary identifier, seeing as emails are unique). On instance of a superadmin, then we dont care what he was trying to do. He gets automatic access.

seyi-adeleke commented on server/utilities/processUrl.js:
```diff
@@ -0,0 +1,17 @@
+import url from 'url';
+
+/**
```
Test this please

victor-nweze-deel replied:
I'll get right on this

victor-nweze-deel replied:
Done

PriscillaSam commented:
This looks great. Well done @ProfJigsaw. 👍

seyi-adeleke commented on server/migrations/20181114070235-update-user-model.js:
```diff
@@ -0,0 +1,14 @@
+export default {
+  up: (queryInterface) => {
+    const query = 'DELETE FROM pg_enum '
```
This is weird lol

victor-nweze-deel replied:
Yeah, somehow I have to manually delete the label from the meta table.

seyi-adeleke commented:
Just checked the uncovered lines. Shouldn't be difficult to test.

victor-nweze-deel commented:
> Just checked the uncovered lines. Shouldn't be difficult to test.

I'm working on covering those lines

victor-nweze-deel committed:
ft(access-control): Role based access control

- Create a permissions config file
- Create a middleware to verify user rights
- Include middleware to a protected route
- Write integration test for the new middleware
